### PR TITLE
Add Multi/Parallel Send

### LINKED ISSUE #3: Add Parallel/Multi Send Feature

### PR DESCRIPTION
**Pull request by samyfodil**

### CODE REVIEW

File: streams/client/client.go [hunk 1]
```diff
@@ -112,56 +122,76 @@ func min(t0 time.Time, t1 time.Time) time.Time {
...
-	now := time.Now()
-	ctx, ctx_close := context.WithDeadline(c.ctx, now.Add(SendToPeerTimeout))
-	defer ctx_close()
-	strmDD, _ := ctx.Deadline()
+func (c *Client) MultiSend(cmdName string, body command.Body, thresh int) (map[peerCore.ID]cr.Response, map[peerCore.ID]error, error) {
+	responses, err := c.send(cmdName, body, thresh)
+	if err != nil {
+		return nil, nil, err
+	}
+	rets := make(map[peerCore.ID]cr.Response)
```
Comment (tafseer-khan) [inline streams/client/client.go:148]: new line before, can make fixed size

File: streams/client/client.go [hunk 2]
```diff
@@ -112,56 +122,76 @@ func min(t0 time.Time, t1 time.Time) time.Time {
...
+	return resp.Response, nil
 }
 
-func (c *Client) send(cmdName string, body command.Body) (peerCore.ID, cr.Response, error) {
-	cmd := command.New(cmdName, body)
-	now := time.Now()
-	ctx, ctx_close := context.WithDeadline(c.ctx, now.Add(SendToPeerTimeout))
-	defer ctx_close()
-	strmDD, _ := ctx.Deadline()
+func (c *Client) MultiSend(cmdName string, body command.Body, thresh int) (map[peerCore.ID]cr.Response, map[peerCore.ID]error, error) {
```
Comment (tafseer-khan) [inline streams/client/client.go:142]: thresh -> threshold

File: streams/client/client.go [hunk 3]
```diff
@@ -112,56 +122,76 @@ func min(t0 time.Time, t1 time.Time) time.Time {
...
+	responses, err := c.send(cmdName, body, thresh)
+	if err != nil {
+		return nil, nil, err
+	}
+	rets := make(map[peerCore.ID]cr.Response)
+	errs := make(map[peerCore.ID]error)
+	for resp := range responses {
+		if resp.error != nil {
+			errs[resp.ID] = resp.error
+		} else {
```
Comment (tafseer-khan) [inline streams/client/client.go:153]: Is there instance where there is error and response?

File: streams/client/client.go [hunk 4]
```diff
@@ -112,56 +122,76 @@ func min(t0 time.Time, t1 time.Time) time.Time {
...
+				continue
+			}
 
-		if slices.Contains(protos, proto) {
-			peers <- peerCore.AddrInfo{ID: peer, Addrs: c.node.Peer().Peerstore().Addrs(peer)}
-		}
-	}
+			protos, err := c.node.Peer().Peerstore().GetProtocols(peer)
+			if err != nil {
+				logger.Errorf("getting protocols for `%s` failed with: %s", peer, err)
```
Comment (tafseer-khan) [inline streams/client/client.go:181]: err.Error(), logger wont parse the error properly

File: streams/client/client.go [hunk 5]
```diff
@@ -112,56 +122,76 @@ func min(t0 time.Time, t1 time.Time) time.Time {
...
+				peers <- peerCore.AddrInfo{ID: peer, Addrs: c.node.Peer().Peerstore().Addrs(peer)}
+			}
 		}
 
-		go func() {
-			defer close(peers)
+		if len(peers) == 0 {
+			discPeers, err := c.node.Discovery().FindPeers(ctx, c.path, discovery.Limit(DiscoveryLimit))
+			if err != nil {
+				logger.Errorf("discovering nodes for `%s` failed with: %s", proto, err)
```
Comment (tafseer-khan) [inline streams/client/client.go]: same

File: streams/client/client.go [hunk 6]
```diff
@@ -178,82 +208,132 @@ func (c *Client) send(cmdName string, body command.Body) (peerCore.ID, cr.Respon
...
-			if err != nil {
-				logger.Errorf("starting stream to `%s`;`%s` failed with: %s", pid.String(), c.path, err)
-				continue
-			}
-		case <-ctx.Done():
-			return "", nil, fmt.Errorf("%s: finding peers for command `%s`.`%s` timed out!: %s", c.node.ID(), c.path, cmdName, ctx.Err())
-		}
+	strm, err := c.node.Peer().NewStream(network.WithNoDial(c.ctx, "application ensured connection to peer exists"), peer.ID, protocol.ID(c.path))
+	if err != nil {
+		logger.Errorf("starting stream to `%s`;`%s` failed with: %s", peer.ID.String(), c.path, err)
```
Comment (tafseer-khan) [inline streams/client/client.go]: same

File: streams/client/client.go [hunk 7]
```diff
@@ -178,82 +208,132 @@ func (c *Client) send(cmdName string, body command.Body) (peerCore.ID, cr.Respon
...
+
+func (c *Client) sendTo(strm stream, deadline time.Time, cmdName string, body command.Body) response {
+	cmd := command.New(cmdName, body)
 
-	if err := strm.SetWriteDeadline(min(time.Now().Add(SendTimeout), strmDD)); err != nil {
-		return "", nil, fmt.Errorf("setting write deadline at `%s`;`%s` failed with: %s", pid.String(), c.path, err)
+	if err := strm.SetWriteDeadline(min(time.Now().Add(SendTimeout), deadline)); err != nil {
+		return response{
+			ID:    strm.ID,
+			error: fmt.Errorf("setting write deadline failed with: %s", err),
```
Comment (tafseer-khan) [inline streams/client/client.go]: %w for error

File: streams/client/client.go [hunk 8]
```diff
@@ -178,82 +208,132 @@ func (c *Client) send(cmdName string, body command.Body) (peerCore.ID, cr.Respon
...
+		}
 	}
 
 	if err := cmd.Encode(strm); err != nil {
-		dl, ok := ctx.Deadline()
-		return "", nil, fmt.Errorf("encoding at `%s`;`%s` t(%s::%v) failed with: %s", pid.String(), c.path, dl.String(), ok, err)
+		return response{
+			ID:    strm.ID,
+			error: fmt.Errorf("seding command `%s(%s)` failed with: %s", cmd.Command, c.path, err),
```
Comment (tafseer-khan) [inline streams/client/client.go]: same

File: streams/client/client.go [hunk 9]
```diff
@@ -178,82 +208,132 @@ func (c *Client) send(cmdName string, body command.Body) (peerCore.ID, cr.Respon
...
+		}
 	}
 
-	if err := strm.SetReadDeadline(min(time.Now().Add(RecvTimeout), strmDD)); err != nil {
-		return "", nil, fmt.Errorf("setting read deadline at `%s`;`%s` failed with: %s", pid.String(), c.path, err)
+	if err := strm.SetReadDeadline(min(time.Now().Add(RecvTimeout), deadline)); err != nil {
+		return response{
+			ID:    strm.ID,
+			error: fmt.Errorf("setting read deadline failed with: %s", err),
```
Comment (tafseer-khan) [inline streams/client/client.go]: same

File: streams/client/client.go [hunk 10]
```diff
@@ -178,82 +208,132 @@ func (c *Client) send(cmdName string, body command.Body) (peerCore.ID, cr.Respon
...
 	}
 
-	response, err := cr.Decode(strm)
+	resp, err := cr.Decode(strm)
 	if err != nil {
-		dl, ok := ctx.Deadline()
-		return "", nil, fmt.Errorf("decoding at `%s`;`%s` t(%s::%v) failed with: %s", pid.String(), c.path, dl.String(), ok, err)
+		return response{
+			ID:    strm.ID,
+			error: fmt.Errorf("recv response of `%s(%s)` failed with: %s", cmd.Command, c.path, err),
```
Comment (tafseer-khan) [inline streams/client/client.go]: same

File: streams/client/client.go [hunk 11]
```diff
@@ -178,82 +208,132 @@ func (c *Client) send(cmdName string, body command.Body) (peerCore.ID, cr.Respon
...
+		}
 	}
 
-	if v, k := response["error"]; k {
-		return pid, cr.Response{}, errors.New(fmt.Sprint(v))
+	if v, k := resp["error"]; k {
+		return response{
+			ID:    strm.ID,
+			error: errors.New(fmt.Sprint(v)),
```
Comment (tafseer-khan) [inline streams/client/client.go:272]: Is this ok? Shouldnt it be safer to fmt.Sprintf("%v",v)

File: streams/client/client.go [hunk 12]
```diff
@@ -178,82 +208,132 @@ func (c *Client) send(cmdName string, body command.Body) (peerCore.ID, cr.Respon
...
+	strmsCount := 0
+	go func() {
+		defer close(strms)
+
+		peers := make(chan peerCore.AddrInfo, 64)
+		defer close(peers)
+
+		for {
+			if strmsCount >= minStreams {
+				return
```
Comment (tafseer-khan) [inline streams/client/client.go:299]: shouldnt we store ctxC variable and call here?

File: streams/client/client_test.go [hunk 13]
```diff
@@ -164,3 +164,133 @@ func TestClient(t *testing.T) {
...
+		t.Errorf("Peer creation returned error `%s`", err.Error())
+		return
+	}
+	defer p1.Close()
+
+	svr1, err := peerService.New(p1, "hello", "/hello/1.0")
+	if err != nil {
+		t.Errorf("Service creation returned error `%s`", err.Error())
+		return
+	}
```
Comment (tafseer-khan) [inline streams/client/client_test.go:201]: new line

File: streams/client/client_test.go [hunk 14]
```diff
@@ -164,3 +164,133 @@ func TestClient(t *testing.T) {
...
+	defer svr1.Stop()
+	err = svr1.Define("hi", func(context.Context, streams.Connection, command.Body) (cr.Response, error) {
```
Comment (tafseer-khan) [inline streams/client/client_test.go:204]: one line error

File: streams/client/client_test.go [hunk 15]
```diff
@@ -164,3 +164,133 @@ func TestClient(t *testing.T) {
...
+		t.Errorf("Peer creation returned error `%s`", err.Error())
+		return
+	}
+	defer p2.Close()
+
+	svr2, err := peerService.New(p2, "hello", "/hello/1.0")
+	if err != nil {
+		t.Errorf("Service creation returned error `%s`", err.Error())
+		return
+	}
```
Comment (tafseer-khan) [inline streams/client/client_test.go:232]: new line

File: streams/client/client_test.go [hunk 16]
```diff
@@ -164,3 +164,133 @@ func TestClient(t *testing.T) {
...
+	defer svr2.Stop()
+	err = svr2.Define("hi", func(context.Context, streams.Connection, command.Body) (cr.Response, error) {
```
Comment (tafseer-khan) [inline streams/client/client_test.go:235]: one line

File: streams/client/client_test.go [hunk 17]
```diff
@@ -164,3 +164,133 @@ func TestClient(t *testing.T) {
...
+		true,
+		false,
+	)
+	if err != nil {
+		t.Errorf("Ping test returned error `%s`", err.Error())
+		return
+	}
+	defer p3.Close()
+
+	err = p3.Peer().Connect(ctx, peercore.AddrInfo{ID: p1.ID(), Addrs: p1.Peer().Addrs()})
```
Comment (tafseer-khan) [inline streams/client/client_test.go:259]: one line

File: streams/client/client_test.go [hunk 18]
```diff
@@ -164,3 +164,133 @@ func TestClient(t *testing.T) {
...
+	if err != nil {
+		t.Errorf("Connect to peer %v returned `%s`", p1.Peer().Addrs(), err.Error())
+		return
+	}
+
+	err = p3.Peer().Connect(ctx, peercore.AddrInfo{ID: p2.ID(), Addrs: p2.Peer().Addrs()})
```
Comment (tafseer-khan) [inline streams/client/client_test.go:265]: same

File: streams/client/client_test.go [hunk 19]
```diff
@@ -164,3 +164,133 @@ func TestClient(t *testing.T) {
...
+		t.Errorf("Connect to peer %v returned `%s`", p2.Peer().Addrs(), err.Error())
+		return
+	}
+
+	// discover
+	cd, err := New(ctx, p3, nil, "/hello/1.0", 2, 2)
+	if err != nil {
+		t.Errorf("Client creation returned error `%s`", err.Error())
+		return
+	} else {
```
Comment (tafseer-khan) [inline streams/client/client_test.go]: done need else

File: streams/client/client_test.go [hunk 20]
```diff
@@ -164,3 +164,133 @@ func TestClient(t *testing.T) {
...
+		}
+
+		for p, r := range res {
+			if r["message"] != "HI" {
+				t.Errorf("node %s returned bad response `%s`", p.Pretty(), r)
+				return
+			}
+		}
+
+		//Close
```
Comment (tafseer-khan) [inline streams/client/client_test.go]: unnecessary comment

File: streams/client/client_test.go [hunk 21]
```diff
@@ -164,3 +164,133 @@ func TestClient(t *testing.T) {
...
+			return
+		}
+
+		if len(res) != 2 && len(errs) == 0 {
+			t.Errorf("MultiSending command failed R=%d, E=%d", len(res), len(errs))
+			return
+		}
+
+		for p, r := range res {
+			if r["message"] != "HI" {
```
Comment (tafseer-khan) [inline streams/client/client_test.go]: if val, ok := 
its just a test, but still should try to avoid panics if something goes wrong

**Review by tafseer-khan — CHANGES_REQUESTED** [review]

File: streams/client/client.go (hunk 3 above)
Comment (samyfodil) [reply streams/client/client.go:153]: yup

File: streams/client/client.go (hunk 11 above)
Comment (samyfodil) [reply streams/client/client.go:272]: Nope.

File: streams/client/client_test.go (hunk 14 above)
Comment (samyfodil) [reply streams/client/client_test.go:204]: will look ugly

File: streams/client/client_test.go (hunk 16 above)
Comment (samyfodil) [reply streams/client/client_test.go:235]: ugly

File: streams/client/client_test.go (hunk 17 above)
Comment (samyfodil) [reply streams/client/client_test.go:259]: same

File: streams/client/client_test.go (hunk 18 above)
Comment (samyfodil) [reply streams/client/client_test.go:265]: could be, but all the other are not

File: streams/client/client_test.go (hunk 21 above)
Comment (samyfodil) [reply streams/client/client_test.go]: if val does not exist in a map the nil value of the val type is returned. here it's ""

**Pull request merged**